### PR TITLE
Add version switch to CLI

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

**Comment by newhoggy** [comment]
Examples:

```
$ bech32 --help
Usage: bech32 [PREFIX | (-v|--version)]
  Convert to and from bech32 strings. Data are read from standard input.

Available options:
  -h,--help                Show this help text
  PREFIX                   output version information and exit

Supported encoding formats: Base16, Bech32 & Base58.

Examples:
  To Bech32:
    $ bech32 base16_ <<< 706174617465
    base16_1wpshgct5v5r5mxh0

    $ bech32 base58_ <<< Ae2tdPwUPEYy
    base58_1p58rejhd9592uusa8pzj2

    $ bech32 new_prefix <<< old_prefix1wpshgcg2s33x3
    new_prefix1wpshgcgeak9mv

  From Bech32:
    $ bech32 <<< base16_1wpshgct5v5r5mxh0
    706174617465
$ bech32 --version
1.1.0
$ bech32 base16_ <<< 706174617465
base16_1wpshgct5v5r5mxh0
```

**Comment by newhoggy** [comment]
This addresses https://github.com/input-output-hk/bech32/issues/23

File: bech32/app/Main.hs [hunk 1]
```diff
@@ -100,19 +109,20 @@ parse = customExecParser (prefs showHelpOnEmpty) parser
         ]
 
     cmd :: Parser Cmd
-    cmd = Cmd <$> optional hrpArgument
+    cmd = (RunCmd <$> optional hrpArgument) <|> (VersionCmd <$ versionSwitch)
```
Comment (rvl) [inline bech32/app/Main.hs:112]: This looks a bit odd.
```
bech32 [PREFIX | (-v|--version)]
```
Can you hide the `--version` flag from the usage text, but still have it shown in the options list?
I _think_ the modifier is [Options.Applicative.hidden](https://hackage.haskell.org/package/optparse-applicative-0.14.2.0/docs/Options-Applicative.html#v:hidden).

File: bech32/app/Main.hs [hunk 2]
```diff
@@ -100,19 +109,20 @@ parse = customExecParser (prefs showHelpOnEmpty) parser
...
+
+versionSwitch :: Parser Bool
```
Comment (rvl) [inline bech32/app/Main.hs:114]: I think that `--version` is actually a flag not a switch. That is, this type should be `Parser ()`, or perhaps `Parser Cmd`.

File: bech32/app/Main.hs [hunk 3]
```diff
@@ -100,19 +109,20 @@ parse = customExecParser (prefs showHelpOnEmpty) parser
...
 hrpArgument = argument (eitherReader reader) $ mconcat
     [ metavar "PREFIX"
     , helpDoc $ Just $ vsep
-        [ text "An optional human-readable prefix (e.g. 'addr')."
-        , indent 2 $ text
-            "- When provided, the input text is decoded from various encoding \
-            \formats and re-encoded to bech32 using the given prefix."
-        , indent 2 $ text
-            "- When omitted, the input text is decoded from bech32 to base16."
+        [ text "output version information and exit"
```
Comment (rvl) [inline bech32/app/Main.hs:125]: What's going on here?

File: bech32/app/Main.hs [hunk 4]
```diff
@@ -48,15 +50,22 @@ import qualified Codec.Binary.Bech32 as Bech32
 import qualified Codec.Binary.Bech32.Internal as Bech32
 import qualified Data.ByteArray.Encoding as BA
 import qualified Data.ByteString.Char8 as B8
+import qualified Data.List as L
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
+import qualified Options.Applicative as OA
```
Comment (rvl) [inline bech32/app/Main.hs:56]: The coding style of this module uses less qualified imports.
I don't think there need to be any more qualified imports than what there currently are.
Even `import Paths_bech32 (version)` is fine.

File: bech32/app/Main.hs [hunk 5]
```diff
@@ -121,11 +131,13 @@ hrpArgument = argument (eitherReader reader) $ mconcat
 
 -- | Run a Command in IO
 run :: Cmd -> IO ()
-run Cmd{prefix} = do
-    source <- T.decodeUtf8 . B8.filter (/= '\n') <$> B8.hGetContents stdin
-    case prefix of
-        Nothing  -> runDecode source
-        Just hrp -> runEncode hrp source
+run cmd = case cmd of
+    VersionCmd -> IO.putStrLn . L.intercalate "." $ fmap show (versionBranch P.version)
```
Comment (rvl) [inline bech32/app/Main.hs:135]: You can just use `Data.Version.showVersion` and `putStrLn` from prelude.

File: bech32/app/Main.hs [hunk 6]
```diff
@@ -100,19 +109,20 @@ parse = customExecParser (prefs showHelpOnEmpty) parser
...
 
     cmd :: Parser Cmd
-    cmd = Cmd <$> optional hrpArgument
+    cmd = (RunCmd <$> optional hrpArgument) <|> (VersionCmd <$ versionSwitch)
+
+versionSwitch :: Parser Bool
+versionSwitch = OA.switch $ mconcat
+    [ OA.long "version"
+    , OA.short 'v'
+    ]
```
Comment (rvl) [inline bech32/app/Main.hs:118]:
```suggestion
    , help "output version information and exit"
    ]
```

File: bech32/app/Main.hs [hunk 7]
```diff
@@ -48,15 +50,22 @@ import qualified Codec.Binary.Bech32 as Bech32
...
 main = setup >> parse >>= run
 
-newtype Cmd = Cmd
-  { prefix :: Maybe HumanReadablePart
-  } deriving (Show)
+data Cmd
+    = RunCmd
+    { prefix :: Maybe HumanReadablePart
+    }
+    | VersionCmd
```
Comment (rvl) [inline bech32/app/Main.hs:67]: Record fields in sum type are totally OK by me for this type of usage.

**Review by rvl — CHANGES_REQUESTED** [review]
Thanks! I was wanting this switch do be added.

True to the parable of the bike shed, I have a few changes requested!

File: bech32/app/Main.hs (hunk 3 above)
Comment (newhoggy) [reply bech32/app/Main.hs]: Oops

**Review by rvl — APPROVED** [review]
Looks fine now, thanks.

I also saw there is a slightly nicer way to do the version option, if you're interesting in trying that: https://www.fpcomplete.com/haskell/library/optparse-applicative/

**Comment by newhoggy** [comment]
> I also saw there is a slightly nicer way to do the version option, if you're interesting in trying that

Maybe later 😁

**Pull request merged**